crates/bevy_ecs/src/system/query.rs

```diff
@@ -9,246 +9,270 @@ use crate::{
-/// over every query result. Refer to the
-/// [`Iterator` API docs](https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html)
-/// for advanced iterator usage.
+/// |Query methods|Effect|
+/// |:---:|---|
+/// |[`iter`]\([`_mut`][`iter_mut`])|Returns an iterator over all query items.|
+/// |[`for_each`]\([`_mut`][`for_each_mut`]),<br>[`par_for_each`]\([`_mut`][`par_for_each_mut`])|Runs a specified function for each query item.|
+/// |[`iter_many`]\([`_mut`][`iter_many_mut`])|Iterates or runs a specified function over query items generated by a list of entities.|
+/// |[`iter_combinations`]\([`_mut`][`iter_combinations_mut`])|Returns an iterator over all combinations of a specified number of query items.|
+/// |[`get`]\([`_mut`][`get_mut`])|Returns the query item for the specified entity.|
+/// |[`many`]\([`_mut`][`many_mut`]),<br>[`get_many`]\([`_mut`][`get_many_mut`])|Returns the query items for the specified entities.|
+/// |[`single`]\([`_mut`][`single_mut`]),<br>[`get_single`]\([`_mut`][`get_single_mut`])|Returns the query item while verifying that there aren't others.|
 ///
-/// ```
-/// # use bevy_ecs::prelude::*;
-/// # #[derive(Component)]
-/// # struct ComponentA;
-/// # #[derive(Component)]
-/// # struct ComponentB;
-/// fn immutable_query_system(query: Query<(&ComponentA, &ComponentB)>) {
-///     for (a, b) in &query {
-///         // Here, `a` and `b` are normal references to components, relatively of
-///         // `&ComponentA` and `&ComponentB` types.
-///     }
-/// }
-/// # bevy_ecs::system::assert_is_system(immutable_query_system);
+/// There are two methods for each type of query operation: immutable and mutable (ending with `_mut`).
+/// When using immutable methods, the query items returned are of type [`ROQueryItem`], a read-only version of the query item.
+/// In this circumstance, every mutable reference in the query fetch type parameter is substituted by a shared reference.
 ///
-/// fn mutable_query_system(mut query: Query<(&mut ComponentA, &ComponentB)>) {
-///     for (mut a, b) in &mut query {
-///         // Similar to the above system, but this time `ComponentA` can be accessed mutably.
-///         // Note the usage of `mut` in the tuple and the call to `iter_mut` instead of `iter`.
-///     }
-/// }
-/// # bevy_ecs::system::assert_is_system(mutable_query_system);
-/// ```
+/// # Performance
+///
+/// Creating a `Query` is a low-cost constant operation.
+/// Iterating it, on the other hand, fetches data from the world and generates items, which can have a significant computational cost.
+///
+/// [`Table`] component storage type is much more optimized for query iteration than [`SparseSet`].
+///
+/// Two systems cannot be executed in parallel if both access the same component type where at least one of the accesses is mutable.
+/// This happens unless the schedule can verify that no entity could be found in both queries.
```

Comment:
```suggestion
/// This happens unless the executor can verify that no entity could be found in both queries.
```
If I understood correctly, in this case is the executor that verifies it, not the schedule itself, right?

I'm using [Stageless RFC](https://github.com/alice-i-cecile/rfcs/blob/stageless/rfcs/45-stageless.md#scheduling-terminology-and-overview) definition:
```
- schedule (noun): the executable form of a system set
- schedule (verb): specify when and under what conditions systems run
- executor: runs the systems in a schedule on a world
```

Reply:
Actually, I don't know how the current schedule works down to this detail. Since the schedule contains the executor, if your suggestion is correct, also is the current version. Ping @alice-i-cecile for this nit. I think we can go either way though.

Reply:
The executor is the agent that actually verifies this, correct.